system_metrics_collector/include/moving_average_statistics/moving_average.hpp

```diff
@@ -0,0 +1,189 @@
+ *  for standard deviation.
+ *
+ *  When statistics are not available, e.g. no observations have been made, NaNs are returned.
+ *
+ *  @tparam T the type to be observed. Note: this must conform to std::is_arithmetic.
+**/
+
+template<
+  class T,
+  typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+class MovingAverageStatistics
+{
+public:
+  /**
+   *  MovingAverageStatistics object constructor.
+  **/
+  MovingAverageStatistics() = default;
+  ~MovingAverageStatistics() = default;
+
+  /**
+   *  Returns the arithmetic mean of all data recorded. If no observations have been made, returns NaN.
+   *
+   *  @return The arithmetic mean of all data recorded, or NaN if the sample count is 0.
+  **/
+  double average()
+  {
+    return getStatistics().average;
+  }
+  /**
+   *  Returns the maximum value recorded. If size of list is zero, returns NaN.
+   *
+   *  @return The maximum value recorded, or NaN if size of data is zero.
+  **/
+  double max()
+  {
+    return getStatistics().max;
+  }
+  /**
+   *  Returns the minimum value recorded. If size of list is zero, returns NaN.
+   *
+   *  @return The minimum value recorded, or NaN if size of data is zero.
+  **/
+  double min()
+  {
+    return getStatistics().min;
+  }
+  /**
+   *  Returns the standard deviation (population) of all data recorded. If size of list is zero, returns NaN.
+   *
+   *  Variance is obtained by Welford's online algorithm,
+   *  see https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford%27s_online_algorithm
+   *
+   *  @return The standard deviation (population) of all data recorded, or NaN if size of data is zero.
+  **/
+  double standardDeviation()
+  {
+    return getStatistics().standard_deviation;
+  }
+  /**
+   *  Return a StatisticResults object, containing average, minimum, maximum, standard deviation (population),
+   *  and sample count.
+   *  For the case of no observations, the average, min, max, and standard deviation are NaN.
+   *
+   *  @return StatisticResults object, containing average, minimum, maximum, standard deviation (population),
+   *  and sample count.
+  **/
+  StatisticResults getStatistics()
+  {
+    std::lock_guard<std::recursive_mutex> guard(mutex);
+    StatisticResults to_return;
+    to_return.sample_count = count_;
+
+    if (count_ == 0) {
+      to_return.average = nan("");
+      to_return.min = nan("");
+      to_return.max = nan("");
+      to_return.standard_deviation = nan("");
+    } else {
+      to_return.average = average_;
+      to_return.min = min_;
+      to_return.max = max_;
+      to_return.standard_deviation = std::sqrt(sum_of_square_diff_from_mean_ / count_);
+    }
+
+    return to_return;
+  }
+  /**
+   *  Reset all calculated values. Equivalent to a new window for a moving average.
+  **/
+  void reset()
+  {
+    std::lock_guard<std::recursive_mutex> guard(mutex);
+    average_ = 0;
+    min_ = DBL_MAX;
+    max_ = DBL_MIN;
```

Comment:
Consider using https://en.cppreference.com/w/cpp/types/numeric_limits instead as it is more portable than those older platform-specific C macros.